rsfbclient-rust/src/client.rs

```diff
@@ -455,7 +459,18 @@ impl FirebirdWireConnection {
         )?)?;
         self.socket.flush()?;
 
-        let (op_code, mut resp) = self.read_packet()?;
+        let (mut op_code, mut resp) = self.read_packet()?;
+
+        // Read lazy responses
+        for _ in 0..self.lazy_count {
+            if op_code != WireOp::Response as u32 {
+                return err_conn_rejected(op_code);
+            }
+            self.lazy_count -= 1;
+            parse_response(&mut resp)?;
+
+            op_code = resp.get_u32()?;
+        }
```

Comment:
@jairinhohw I think you probably wrote this,

I think I understand the idea, but maybe you could elaborate a little bit the on 'lazy response' changes?

1. What's the intention? Is it to batch syscalls as much as possible and defer blocking?
2. Is wire protocol response order guaranteed to match TCP order of requests (if so- is that written somewhere or based just on observations?)
3. (followup to (2)) is there any other facility like a 'correlation id' to match req-resp pairs perfectly? it does not seem so.

any other details you think might be enlightening are also appreciated

Reply:
The link in the mailing list is indeed about the feature that makes it necessary to add those lazy responses verifications.

Some requests in the protocol version 11 and above do not cause the server to send the response immediately, the response is only sent after another request is made and needs a immediate response (does not have a deferred response), so all pending responses are sent from the server in the order of the requests, ending with the response to the last request the client made.

I've commented in the code the requests that causes deferred responses and incremented the `lazy_count` field to track how many there are, and in all parts that receive a server response and may have some pending deferred response to receive, I've added that check to receive any pending response.

About the order of the responses, at least up to protocol version 13 (i don't know the ones above this), there is no identifier to tie a request to a response, so all responses are sent in the same order of the requests.

Without this change, there was a chance that a response from a `free_statement` request was read in place of a response from another request, and the response of that request would be ignored.

This behavior is commented [here](https://firebirdsql.org/file/documentation/html/en/firebirddocs/wireprotocol/firebird-wire-protocol.html#wireprotocol-statements-free-v11).